lib/selections.js

```diff
@@ -0,0 +1,222 @@
+/**
+ * @typedef {Object} MinimalSelectionItem
+ * @property {number} from
+ * @property {number} to
+ */
+
+/** A selection of pieces to download.
+ * @typedef {MinimalSelectionItem & {
+ *  offset: number,
+ *  priority: number,
+ *  isStreamSelection?: boolean
+ * }} SelectionItem
+ */
+
+/**
+ * @typedef {MinimalSelectionItem & {notify: function}} NotificationItem
+ */
+
+export class Selections {
+  /** @type {Array<SelectionItem>} */
+  _items = []
+
+  /** @type {Array<NotificationItem>} */
+  _notifications = []
+
+  /**
+   * @param {MinimalSelectionItem & {isStreamSelection: boolean}} item Interval to be removed from the selection
+   */
+  remove (item) {
+    for (let i = 0; i < this._items.length; i++) {
+      const existing = this._items[i]
+
+      if (existing.isStreamSelection) {
+        if (item.isStreamSelection) {
+          // If both are stream selections and they match, then we remove the first matching item, then we break the loop
+          if (existing.from === item.from && existing.to === item.to) {
+            this._items.splice(i, 1)
+            // for stream selections, we only remove one item at a time
+            // ergo we break the loop after removing the first matching item
+            break
+          }
+        } else {
+          // we only remove stream selections when the `isStreamSelection` flag is true and they match
+          continue
+        }
+      }
+
+      if (isLowerIntersecting(item, existing)) {
+        existing.to = Math.max(item.from - 1, 0)
+        continue
+      }
+      if (isUpperIntersecting(item, existing)) {
+        existing.from = item.to + 1
+        continue
+      }
+      if (isInsideExisting(item, existing)) {
+        const existingStart = { ...existing, to: Math.max(item.from - 1, 0) }
+        const existingEnd = { ...existing, from: item.to + 1 }
+        if (existingStart.to - existingStart.from > 0) {
+          this._items.splice(i, 1, existingStart, existingEnd)
+        } else {
+          this._items.splice(i, 1, existingEnd)
+        }
+        continue
+      }
+      if (isCoveringExisting(item, existing)) {
+        this._items.splice(i, 1)
+        i -= 1
+        continue
+      }
+    }
+  }
+
+  /**
+   * @param {SelectionItem} newItem
+   */
+  insert (newItem) {
+    const { from, to, notify, isStreamSelection, ...restItem } = newItem
+    if (from > to) {
+      throw new Error('Invalid interval')
+    }
+    if (!isStreamSelection) {
+      this.remove({ from, to, isStreamSelection })
+    }
+    this._items.push({ from, to, isStreamSelection, ...restItem })
+    this._notifications.push({ from, to, notify })
```

Comment:
I don't like this, you're essentially using the old range behavior, and it has it's own separate state, not like the rest, it's way too complex